Introduce `RUSTUP_DOWNLOAD_TIMEOUT` to override the download timeout

FranciscoTGouveia opened the pull request:
Fixes #4439.

Besides increasing the default timeout from 30 seconds to 3 minutes, a new environment variable (`RUSTUP_DOWNLOAD_TIMEOUT`) is introduced to allow users to override this value.

rami3l commented on src/download/mod.rs:
```diff
@@ -322,9 +326,15 @@ impl Backend {
+        let timeout = Duration::from_secs(match process.var("RUSTUP_DOWNLOAD_TIMEOUT") {
+            Ok(s) => s.parse::<NonZeroU64>().context(
+                "invalid value in RUSTUP_DOWNLOAD_TIMEOUT -- must be a natural number greater than zero",
+            ).unwrap().get(),
```
Why are we using an `.unwrap()` here?

FranciscoTGouveia replied:
To get the `NonZeroU64` out of the `Result`; but after reading about good practices, I realized that using `?` instead of `.unwrap()` is preferable.
Is this what you were referring to?

djc replied:
FYI, IMO it's nicer to write `NonZeroU64::from_str(s)` compared to `s.parse::<NonZeroU64>()`.

FranciscoTGouveia replied:
Thank you for the suggestion, this way I can avoid the *turbofish*.
However, doesn’t this require an additional import (`use std::str::FromStr`)?
Would it be acceptable to include the extra import in this case?

djc replied:
Yes, just add the extra import at the top of the module with the other `std` ones.

djc commented on src/download/mod.rs:
```diff
@@ -622,21 +634,24 @@ mod reqwest_be {
+    }
+
+    #[cfg(feature = "reqwest-native-tls")]
+    static CLIENT_NATIVE_TLS: OnceLock<Client> = OnceLock::new();
```
The change from `LazyLock` to `OnceLock` seems unrelated and doesn't look like an improvement to me.

FranciscoTGouveia replied:
Thank you for pointing this out!
The motivation for this change was to allow me to pass an argument upon the creation of the *client*.
Since `LazyLock` does not allow passing any arguments (the `timeout`), I decided to switch to an `OnceLock`.
Is there another way to achieve this that I might be overlooking?

djc replied:
Ah yes, I guess that makes sense -- sorry for missing that. I think this would be easier to review if you extract the move from `LazyLock` to `OnceLock` into a separate commit.

FranciscoTGouveia replied:
I will now divide my changes in two separate commits.

djc reviewed (commented):
Nice to see that you are now passing `Duration` by value!

rami3l commented on doc/user-guide/src/environment-variables.md:
```diff
@@ -66,6 +66,9 @@
 
 - `RUSTUP_TERM_WIDTH` (default: none). Allows to override the terminal width for progress bars.
 
+- `RUSTUP_DOWNLOAD_TIMEOUT` (default: 180). Allows to override the default
```
Nit: You probably need to mark this variable as unstable.

FranciscoTGouveia committed:
fix(downloads): substitute the LazyLock for a OnceLock

This change is needed to allow us to pass an argument for the creation
of the client, as it is being done in `rustls_client()`.

FranciscoTGouveia committed:
fix(download_timeout): introduce RUSTUP_DOWNLOAD_TIMEOUT for overriding download timeout

As some (corporate) firewalls take too much time to verify the
downloads, the timeout is reached and the user cannot complete its
installation.

Besides changing the timeout from 30 secs to 3 mins, a new environment
variable is introduced to allow users to override the timeout value to
their taste.

rami3l commented on src/download/mod.rs:
```diff
@@ -620,12 +635,13 @@ mod reqwest_be {
 
     #[cfg(feature = "reqwest-native-tls")]
-    pub(super) fn native_tls_client() -> Result<&'static Client, DownloadError> {
+    pub(super) fn native_tls_client(timeout: Duration) -> Result<&'static Client, DownloadError> {
```
It might be worth commenting that the timeout being passed in does nothing when `CLIENT_NATIVE_TLS` is already initialized (idem for `CLIENT_RUSTLS_TLS`)?

FranciscoTGouveia replied:
Yes, I agree -- the early check may execute and the `timeout` is never utilized.
However, looking at the uses of `CLIENT_NATIVE_TLS`, I do not think that there is any other place where it is initialized -- rendering the check useless.

Am I overlooking something here?

Anyways, I have already added a comment. Thank you.

rami3l replied:
@FranciscoTGouveia You are right, but I believe when people read this code out of context, local reasoning could be tricky as they don't typically take all existing callers into account; rather they might think of all possible callers.

djc commented on src/download/mod.rs:
```diff
@@ -622,21 +620,23 @@ mod reqwest_be {
+            .map_err(DownloadError::Reqwest)?;
+
+        let _ = CLIENT_NATIVE_TLS.set(client);
+
```
Add a comment here similar to what's in `rustls_client()`?

FranciscoTGouveia replied:
I’ve just updated my branch, but it was too late, as the PR had already been merged.
How should I proceed?

djc replied:
Just submit a new PR with the change, please.

FranciscoTGouveia replied:
Is it acceptable for the PR to contain only the comment?
